lib/puppet/application/device.rb

```diff
@@ -308,7 +308,8 @@ def main
             0
           elsif options[:apply]
             # ensure we have a cache folder structure exists for the device
-            FileUtils.mkdir_p(Puppet[:statedir]) unless File.directory?(Puppet[:statedir])
+            subdir = File.join(Puppet[:statedir], 'graphs')
+            FileUtils.mkdir_p(subdir) unless File.directory?(subdir)
```

Comment:
The `statedir` will be created when applying the `:main` settings. Probably better to move the `settings.use` call like so:

```ruby
# this will reload and recompute default settings and create the devices sub vardir
Puppet.settings.use :main, :agent, :ssl

unless options[:resource] || options[:facts] || options[:apply] || options[:libdir]
  # ask for a ssl cert if needed, but at least
  # setup the ssl system for this device.
  ssl_host = setup_host(device.name)
...
```
It also will avoid playing wack-a-mole for other files/directories in the statedir

Reply:
dang - for some reason I thought that has unwanted side-effects. Of course, now I can't repro it anymore :-/